src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,503 @@
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available.
+         * <p>
+         * An implementation should concatenate the salt into a single value
+         * once all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(SecretKey salt) {
+            if (salt != null) {
+                salts.add(salt);
+            } else {
+                throw new NullPointerException("salt must not be null");
+            }
+            return this;
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available. An
+         * implementation should concatenate the salt into a single value once
+         * all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            if (salt == null) {
+                throw new NullPointerException(
+                    "salt must not be null");
+            }
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and
+     * {@code ExtractThenExpand} objects.
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should
+     * be called next, before calling build methods, such as
+     * {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder ofExtract() {
+        return new Builder().createBuilder();
+    }
+
+    /**
+     * Defines the input parameters of an {@code Expand} object
+     *
+     * @param prk
+     *     the pseudorandom key; must not be {@code null} in the Expand case
+     * @param info
+     *     the optional context and application specific information (may be
+     *     {@code null}); the byte[] is copied to prevent subsequent
+     *     modification
+     * @param length
+     *     the length of the output key material (must be &gt; 0 and &lt; 255 *
+     *     HMAC length)
```

Comment:
The maximum length can only be checked in the implementation and not here. Maybe do not mention it.

Reply:
OK. As long as the exception does not cover it.

Reply:
You should be careful putting the word "must" in an API specification - I think this would indicate that it is a condition that needs to be tested. And I think an HKDF implementation should throw an exception if the length is too high, right? I also think there is likely to cause confusion about what the HMAC length is and why the max is 255 * it. Also should it be hash length?

This could be a useful tip, but I also think there it is likely to inevitably prompt questions from users about what the HMAC length is for a given algorithm and why the max is 255 * that; in other words it requires some knowledge of the RFC and crypto. 

Suggest something like:

"the length of the output key material. The length must be > 0 and < 255 * hash length. For example, for the HKDFWithSHA256 algorithm, the hash length is ... so the maximum length is ... If less than 0, this method throws an IllegalArgumentException. If the length exceeds the maximum this method will not throw an exception because the hash algorithm is not known at this time. For example, for the HKDFWithSHA256 algorithm, the hash length is ... so the maximum length is ...  However an HKDF implementation of `KDF` will throw an `InvalidAlgorithmParameterException` when any of the `deriveKey` or `deriveObject` methods are called with a length that exceeds the maximum."

Also, there are other methods like `Builder.thenExpand` that take a length argument that do not contain this wording. They should be consistent.